### PR TITLE
`ultralytics 8.0.128` FastSAM autodownload and super() init

### PR DESCRIPTION
**Pull request by glenn-jocher**
### Summary
🚀🎨📝

This pull request introduces the FastSAM model, a variant of the YOLO model that uses CLIP to generate semantic masks for object detection. It adds and modifies several files in the `ultralytics/yolo/fastsam` directory, as well as the documentation and the download utility. It also improves the code quality, style, and documentation of the existing files.

> _Oh we are the coders of the FastSAM_
> _We inherit from YOLO and we don't give a damn_
> _We download the weights and we plot the masks_
> _And we clip the prompts with a bit of class_

### Walkthrough
*  Inherit `FastSAM` class from `YOLO` class and update default model name to `FastSAM-x.pt` ([link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-e903ab4ea12ad024cb11630009bc70836cd9429887e61b0301619092d2613e54L17-R27))
*  Change image path for FastSAM prediction example to show more diverse and challenging scene ([link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-a515bc5f1b618c3826864089936dbbc3d01d258fa9bdeab644927979dae4a50fL47-R56))
*  Add license header to `predict.py`, `prompt.py`, and `utils.py` files ([link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-fd7934273a4b88897f746c09649d169f9dc315332a5120ddb8d27ed3cfe90f0eR1-R2), [link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66R1-R2), [link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-252193c694a239794e540442fff1de52078085bd06173f175757430033048aacL1-R17))
*  Move `clip` module import to `__init__` method of `FastSAMPrompt` class and handle ImportError ([link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66L9-R11), [link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66L28-R31))
*  Make `_segment_image` and `_format_results` methods static and simplify annotation dictionary creation ([link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66L28-R31), [link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66L42-R57))
*  Simplify `filter_masks` method by combining if conditions, using `extend` method, and normalizing color values ([link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66L62-R76))
*  Correct spelling of macOS and adjust indentation of arguments in `prompt.py` file ([link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66L108-R110), [link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66L167-R171), [link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66L192-R202), [link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66L248-R258))
*  Normalize color values and round and clip bounding box coordinates in `prompt.py` file ([link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66L215-R216), [link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66L270-R272), [link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66L355-R359), [link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9a11576da855120cea2de78c92110d97b0d83ae42fd2afae21003b178310bd66L383-R384))
*  Change docstring of `bbox_iou` function to use triple quotes and Google style in `utils.py` file ([link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-252193c694a239794e540442fff1de52078085bd06173f175757430033048aacL28-R41))
*  Simplify if-else block for returning raw iou output in `utils.py` file ([link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-252193c694a239794e540442fff1de52078085bd06173f175757430033048aacL56-R65))
*  Add FastSAM model name to GitHub asset names in `downloads.py` file ([link](https://github.com/ultralytics/ultralytics/pull/3552/files?diff=unified&w=0#diff-9bc1504ab4553f31c8ad4205083c67ac4fac833eea921bdea0c2dcee0bd57c03R23))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements to the FastSAM documentation, model initialization, and related utilities for object detection and segmentation.

### 📊 Key Changes
- Updated the example image path and inference device in the FastSAM documentation.
- Introduced default FastSAM model change from `FastSAM.pt` to `FastSAM-x.pt`.
- Refactored the `predict` and `val` examples in the documentation to reflect the default model updates.
- Included additional model versions (`FastSAM-s.pt` and `FastSAM-x.pt`) in utility scripts for downloads.
- Improved code quality through refactoring; this includes using static methods and optimizing conditional logic.
- Organized imports and removed unnecessary code.
- Adjusted model version number for `ultralytics` package.

### 🎯 Purpose & Impact
- 💡 **Clarity:** Updating documentation improves clarity on how to use the FastSAM model.
- ✅ **Compatibility:** The default model change ensures the latest recommended model is used.
- 🔄 **Refactoring:** Code cleanup enhances maintainability and readability.
- 🛠️ **Utility:** Streamlined download scripts accommodate new FastSAM model variants.
- 🤖 **Version Tracking:** Increment in package version aids in tracking changes and facilitates smoother updates for users.
  
These changes collectively work towards a more robust, user-friendly, and up-to-date FastSAM tool within the Ultralytics ecosystem. Users should benefit from clearer guidance and an enhanced toolkit for their object detection and segmentation tasks.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@berry-ding @AyushExel @fcakyon @Laughing-q this PR cleans up the FastSAM PR a bit, adds model autodownload for `FastSAM-s.pt`, `FastSAM-x.pt` and renames `FastSAM.pt` requests to `FastSAM-x.pt`. It also adds a super init to the model class and defaults the model argument to `model=FastSAM-x.pt`, so you can create a FastSAM model without any additional args, i.e.:

```python
from ultralytics import FastSAM

model = FastSAM()  # will autodownload `FastSAM-x.pt` if no argument passed
```

**Comment by glenn-jocher** [comment]
@berry-ding I think these changes should help the FastSAM usage a bit, but let me know if I accidentally broke anything or modified anything I shouldn't have. Thanks a lot for your contributions!

**Comment by AyushExel** [comment]
Looks great!

**Comment by berry-ding** [comment]
Hi @glenn-jocher , there is a bug in several cuda devices.https://github.com/ultralytics/ultralytics/blob/ad99246ff18828273bec2961c6c0a019173fc718/ultralytics/yolo/fastsam/utils.py#L23C1-L26C81

The code below can fix it

```
    boxes[:, 0] = torch.where(boxes[:, 0] < threshold, torch.tensor(0,dtype=torch.float,device=boxes.device), boxes[:, 0])  # x1
    boxes[:, 1] = torch.where(boxes[:, 1] < threshold, torch.tensor(0,dtype=torch.float,device=boxes.device), boxes[:, 1])  # y1
    boxes[:, 2] = torch.where(boxes[:, 2] > w - threshold, torch.tensor(w,dtype=torch.float,device=boxes.device), boxes[:, 2])  # x2
    boxes[:, 3] = torch.where(boxes[:, 3] > h - threshold, torch.tensor(h,dtype=torch.float,device=boxes.device), boxes[:, 3])  # y2

```

**Comment by glenn-jocher** [comment]
@berry-ding got it! Would it help to add FastSAM to the CI? Though the CI checks are only on CPU github runners...

I'll add this fix in a PR.

**Comment by glenn-jocher** [comment]
@berry-ding I implemented a slightly different fix in https://github.com/ultralytics/ultralytics/pull/3587/commits/921345b17c5512d1022e1b83a16bda642c6ee6ac. Tested in Colab with no errors:

```python
import torch

h, w = 640, 320

threshold = 20

boxes = torch.randn(1000, 4).cuda() * 600

    # Adjust boxes
boxes[boxes[:, 0] < threshold, 0] = 0  # x1
boxes[boxes[:, 1] < threshold, 1] = 0  # y1
boxes[boxes[:, 2] > w - threshold, 2] = w  # x2
boxes[boxes[:, 3] > h - threshold, 3] = h  # y2
```

This will be published in `ultralytics 8.0.131` soon. Can you please verify this fix works correctly with CUDA? Thanks!
